### PR TITLE
Dot scale attention compatible with dtensor

### PR DESCRIPTION
**Pull request by buildwithsuhana**
## Description
This PR implements DTensor compatibility for the dot_product_attention operation in the PyTorch backend. 

Context:
  PyTorch's scaled_dot_product_attention (SDPA) and its flash attention variants often require explicit handling when working with sharded tensors (DTensor).

Key Changes:
   - `keras/src/backend/torch/nn.py`: Added logic to dot_product_attention to detect and handle DTensor inputs for query, key, value, and mask.
   - `keras/src/ops/nn_test.py`: Added test_dot_product_attention_dtensor which verifies the integration in a distributed (simulated) environment.

## Contributor Agreement

**Please review our [AI-Assisted Contribution Policy](../CONTRIBUTING.md#ai-assisted-contribution-policy) and check all boxes below before submitting your PR for review:**

- [x] I am a human, and not a bot.
- [x] I will be responsible for responding to review comments in a timely manner.
- [x] I will work with the maintainers to push this PR forward until submission.

**Note:** Failing to adhere to this agreement may result in your future PRs no longer being reviewed.

### CODE REVIEW

File: keras/src/ops/nn_test.py [hunk 1]
```diff
@@ -2867,6 +2874,82 @@ def test_dot_product_attention(
             outputs, expected, atol=1e-3 if flash_attention else 1e-6
         )
 
+    @pytest.mark.skipif(backend.backend() != "torch", reason="Torch only")
+    def test_dot_product_attention_dtensor(self):
```
Comment (hertschuh) [inline keras/src/ops/nn_test.py]: Please use the pattern added in https://github.com/keras-team/keras/pull/23590

- Extend `TorchDistributedTestMixin`
- `@pytest.mark.no_pytest_xdist`

File: keras/src/backend/torch/nn.py [hunk 2]
```diff
@@ -1545,6 +1547,13 @@ def dot_product_attention(
                     (q_len, kv_len), dtype=torch.bool, device=mask.device
                 )
             )
+            if isinstance(mask, DTensor):
+                causal_mask = DTensor.from_local(
+                    causal_mask,
+                    mask.device_mesh,
+                    [Replicate()] * mask.device_mesh.ndim,
+                )
+                causal_mask = causal_mask.to(mask.to_local().device)
```
Comment (hertschuh) [inline keras/src/backend/torch/nn.py]: I don't understand this last part:
```python
causal_mask = causal_mask.to(mask.to_local().device)
```
`causal_mask` was local, then you distributed it. But then, you're moving it back to the local device.

I guess my questions are:
- why not just `causal_mask = causal_mask.to_local()` like the other ones line 1581-1585
- is there an easier way to get the local slice of the causal mask instead of this complicated round trip?

**Review by hertschuh — CHANGES_REQUESTED** [review]

File: keras/src/ops/nn_test.py (hunk 1 above)
Comment (buildwithsuhana) [reply keras/src/ops/nn_test.py]: Done

File: keras/src/backend/torch/nn.py (hunk 2 above)
Comment (buildwithsuhana) [reply keras/src/backend/torch/nn.py]: I've removed the redundant .to() call and updated the logic to let DTensor sharding automatically handle the local slicing.

**Review by hertschuh — APPROVED** [review]
Thanks!
